src/audio/host.c

```diff
@@ -350,6 +351,14 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
 	if (!hd->no_stream_position) {
 		hd->report_pos += bytes;
 
+		/* update position every period to support host query to
+		 * get the latest value if cont_update_posn == 1.
+		 */
+		if (hd->cont_update_posn == 1) {
+			pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+			mailbox_stream_write(dev->pipeline->posn_offset,
+					     &hd->posn, sizeof(hd->posn));
+		}
```

Comment:
I am a bit confused now on the relationship between 'no_stream_position' and 'cont_update_posn'.

Is there really a hierarchy between the two, or could 'cont_update_posn' be enabled when we don't send any IPC?

Reply:
@plbossart  good question. from the description of no_stream_position is "1 means don't send stream position".  In original design, IPC should be sent when no_stream_position is zero and satisfy the condition below, 
```c
 "if (hd->host_period_bytes != 0 &&
		    hd->report_pos >= hd->host_period_bytes)"
```
if no_stream_position is set as 1, that means host won't get stream position from DSP firmware. Maybe cont_update_posn will be no used? Thanks.

Reply:
I wonder if there is any case that needs to update stream position from DSP firmware without IPC requirement because DSP send IPC : SOF_IPC_STREAM_POSITION to notify kernel then call ipc_period_elapsed to update stream position and call snd_sof_pcm_period_elapsed for period wakeup mode.

Reply:
The position can be available to host in 4 ways today.
1) Host sends IPC (latency adds some variability here - not very accurate)
2) FW sends notifications to host (more accurate than 1 since its a one way message, but still subject to host latencies when receiving the notification).
3) FW writes to shared SRAM for every copy(). Accurate to 1ms (LL scheduling tick)
4) Host reads from HW position registers (Intel HW can do this today).

Reply:
@lgirdwood  Thanks for detail explanation. I have some questions below,
Q1. Does any platform use method 1 since it is not accurate?
Q2. Does Method 3 exist in current design or it is the same as this PR do?
Q3. Can Method 3 be used along? I think we need one way to notify Host to get position?
Thanks a lot.

Reply:
@plbossart Could you please tell me if you still have any concern or suggestion on this PR? Thanks.

Reply:
I still have the concern on what happens in the no_stream_position case. @ujfalusi had a similar question.

In addition, it's not a good idea to have multiple requests to get timestamps. I would use something like this to make sure the timestamp queries are unified.

````diff
diff --git a/src/audio/host.c b/src/audio/host.c
index 8f0495723..f9014b847 100644
--- a/src/audio/host.c
+++ b/src/audio/host.c
@@ -306,6 +306,8 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
        struct host_data *hd = comp_get_drvdata(dev);
        struct comp_buffer *source;
        struct comp_buffer *sink;
+       bool update_mailbox = false;
+       bool send_ipc = false;
        int ret;
 
 
@@ -346,10 +348,12 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
                hd->local_pos = 0;
 #endif
 
+       hd->report_pos += bytes;
+       if (hd->cont_update_posn)
+               update_mailbox = true;
+
        /* Don't send stream position if no_stream_position == 1 */
        if (!hd->no_stream_position) {
-               hd->report_pos += bytes;
-
                /* host_period_bytes is set to zero to disable position update
                 * by IPC for FW version before 3.11, so send IPC message to
                 * driver according to this condition and report_pos.
@@ -361,12 +365,19 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
                        /* send timestamped position to host
                         * (updates position first, by calling ops.position())
                         */
-                       pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
-                       mailbox_stream_write(dev->pipeline->posn_offset,
-                                            &hd->posn, sizeof(hd->posn));
-                       ipc_msg_send(hd->msg, &hd->posn, false);
+                       update_mailbox = true;
+                       send_ipc = true;
                }
        }
+
+       if (update_mailbox) {
+               pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+               mailbox_stream_write(dev->pipeline->posn_offset,
+                                    &hd->posn, sizeof(hd->posn));
+               if (send_ipc)
+                       ipc_msg_send(hd->msg, &hd->posn, false);
+       }
+
 }
 
 /* The host memory is not guaranteed to be continuous and also not guaranteed
````

Reply:
@plbossart  Thanks for providing suggestion to ensure timestamp query are unified. But I am not sure if it is fine to move  the code below to the no_stream_position case(outside of the if(!hd->no_stream_position)) 
```` diff
+       hd->report_pos += bytes;
+       if (hd->cont_update_posn)
+               update_mailbox = true;
````
 because hd->report_pos  is circular and will be reset as zero under the condition. thanks.
```c
		if (hd->host_period_bytes != 0 &&
		    hd->report_pos >= hd->host_period_bytes) {
			hd->report_pos = 0;
```

Reply:
I can't figure out this code, it's just convoluted for no good reason. @lgirdwood can you take a look?

Reply:
@lgirdwood Sorry to bother. Could you please help to check it and give suggestion if any? Thanks.

Reply:
@yaochunhung apologies, I did not see the question.
It looks like to me that we only need the 
```diff
+       if (update_mailbox) {
+               pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+               mailbox_stream_write(dev->pipeline->posn_offset,
+                                    &hd->posn, sizeof(hd->posn));
```
We can write to the mailbox on every pipeline time tick. However, this should be a topology option "update mailbox position" flag (just like the no position update flag). Would this work for you to read from mailbox ? This would mean no IPC traffic and host could read when it pleased ?

Reply:
@lgirdwood  Thanks for comment. I provide this patch and it looks like fine for me. I will provide a update for review. Thanks.